### PR TITLE
Make tile element insertion type explicit

### PR DESCRIPTION
**Pull request by ZehMatt**
Tile elements should not be inserted without a type, this also fixes one issue where it attempts to fix some surface elements but never and calls AsSurface but the type was not set so it would result in nullptr and crash the game.

This is based on #13975 so I will rebase this branch later.

### CODE REVIEW

File: src/openrct2/world/Map.cpp [hunk 1]
```diff
@@ -1184,7 +1153,7 @@ TileElement* tile_element_insert(const CoordsXYZ& loc, int32_t occupiedQuadrants
 
     // Insert new map element
     insertedElement = newTileElement;
-    newTileElement->type = 0;
+    newTileElement->SetType(static_cast<uint8_t>(type));
```
Comment (Gymnasiast) [inline src/openrct2/world/Map.cpp:1156]: This is not entirely the same, I think. The `type` field contains more information besides just the type and now not everything is 0-initialised.

File: src/openrct2/world/TileInspector.cpp [hunk 2]
```diff
@@ -382,7 +382,7 @@ GameActionResultPtr tile_inspector_paste_element_at(const CoordsXY& loc, TileEle
 
         // The occupiedQuadrants will be automatically set when the element is copied over, so it's not necessary to set them
         // correctly _here_.
-        TileElement* const pastedElement = tile_element_insert({ loc, element.GetBaseZ() }, 0b0000);
+        TileElement* const pastedElement = tile_element_insert({ loc, element.GetBaseZ() }, 0b0000, TileElementType::Corrupt);
```
Comment (duncanspumpkin) [inline src/openrct2/world/TileInspector.cpp:385]: I don't think you should use Corrupt here. Use anything but corrupt as we will eventually remove corrupt. Eitherway its overwritten on line 388

File: src/openrct2/world/Map.cpp (hunk 1 above)
Comment (ZehMatt) [reply src/openrct2/world/Map.cpp:1156]: Good point, I'll  initialize type to 0 and call SetType afterwards.

File: src/openrct2/Game.cpp [hunk 3]
```diff
@@ -481,13 +481,13 @@ void game_fix_save_vars()
             if (surfaceElement == nullptr)
             {
                 log_error("Null map element at x = %d and y = %d. Fixing...", x, y);
-                auto tileElement = tile_element_insert(TileCoordsXYZ{ x, y, 14 }.ToCoordsXYZ(), 0b0000);
+                auto tileElement = TileElementInsert<SurfaceElement>(TileCoordsXYZ{ x, y, 14 }.ToCoordsXYZ(), 0b0000);
```
Comment (duncanspumpkin) [inline src/openrct2/Game.cpp:484]: No need for this temporary variable

**Review by duncanspumpkin — APPROVED** [review]

File: src/openrct2/actions/WallPlaceAction.cpp [hunk 4]
```diff
@@ -427,12 +419,11 @@ GameActions::Result::Ptr WallPlaceAction::Execute() const
...
-    if (GetFlags() & GAME_COMMAND_FLAG_GHOST)
-    {
-        wallElement->SetGhost(true);
-    }
+    wallElement->SetGhost(GetFlags() & GAME_COMMAND_FLAG_GHOST);
 
-    res->tileElement = tileElement;
+    res->tileElement = wallElement->as<TileElement>();
+
+    map_animation_create(MAP_ANIMATION_TYPE_WALL, targetLoc);
```
Comment (Gymnasiast) [inline src/openrct2/actions/WallPlaceAction.cpp:426]: Any reason you moved this line?

Comment (ZehMatt) [reply src/openrct2/actions/WallPlaceAction.cpp:426]: It looked a bit out of place between the creation and initialization of the tile element.

**Review by Gymnasiast — APPROVED** [review]

**Pull request merged**